### PR TITLE
bpf: don't sign-extend IPv6 address components

### LINKED ISSUE #27898: ipv6 endpoint connectivity regression 1.13.6 -> 1.15.0-pre.0
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

On v1.13.6, all `cilium connectivity test` checks pass, including IPv6 with this helm config: https://github.com/ocf/kubernetes/blob/c8014d3fd2cf5d0be7a6892f475cd47f59be2b45/apps/cilium.py

<details>
<summary>1.13.6 connectivity test report</summary>

```
ℹ️  Cilium version: 1.13.6
🏃 Running tests...
[=] Test [no-policies]
......................................
[=] Test [no-policies-extra]
................
[=] Test [allow-all-except-world]
..................
[=] Test [client-ingress]
..
[=] Test [client-ingress-knp]
..
[=] Test [allow-all-with-metrics-check]
........
[=] Test [all-ingress-deny]
........
[=] Test [all-ingress-deny-knp]
........
[=] Test [all-egress-deny]
................
[=] Test [all-egress-deny-knp]
................
[=] Test [all-entities-deny]
........
[=] Test [cluster-entity]
..
[=] Test [host-entity]
........
[=] Test [echo-ingress]
....
[=] Test [echo-ingress-knp]
....
[=] Test [client-ingress-icmp]
..
[=] Test [client-egress]
....
[=] Test [client-egress-knp]
....
[=] Test [client-egress-expression]
....
[=] Test [client-egress-expression-knp]
....
[=] Test [client-with-service-account-egress-to-echo]
....
[=] Test [client-egress-to-echo-service-account]
....
[=] Test [to-entities-world]
......
[=] Test [to-cidr-external]
....
[=] Test [to-cidr-external-knp]
....
[=] Test [echo-ingress-from-other-client-deny]
......
[=] Test [client-ingress-from-other-client-icmp-deny]
......
[=] Test [client-egress-to-echo-deny]
......
[=] Test [client-ingress-to-echo-named-port-deny]
....
[=] Test [client-egress-to-echo-expression-deny]
....
[=] Test [client-with-service-account-egress-to-echo-deny]
....
[=] Test [client-egress-to-echo-service-account-deny]
..
[=] Test [client-egress-to-cidr-deny]
....
[=] Test [client-egress-to-cidr-deny-default]
....
[=] Test [health]
....

[=] Skipping Test [north-south-loadbalancing]

[=] Skipping Test [pod-to-pod-encryption]

[=] Skipping Test [node-to-node-encryption]

[=] Skipping Test [north-south-loadbalancing-with-l7-policy]
[=] Test [echo-ingress-l7]
............
[=] Test [echo-ingress-l7-named-port]
............
[=] Test [client-egress-l7-method]
............
[=] Test [client-egress-l7]
..........
[=] Test [client-egress-l7-named-port]
..........

[=] Skipping Test [client-egress-l7-tls-deny-without-headers]

[=] Skipping Test [client-egress-l7-tls-headers]

[=] Skipping Test [client-egress-l7-set-header]

[=] Skipping Test [echo-ingress-auth-always-fail]

[=] Skipping Test [echo-ingress-mutual-auth-spiffe]

[=] Skipping Test [pod-to-ingress-service]

[=] Skipping Test [pod-to-ingress-service-deny-all]

[=] Skipping Test [pod-to-ingress-service-allow-ingress-identity]
[=] Test [dns-only]
..........
[=] Test [to-fqdns]
........

✅ All 42 tests (316 actions) successful, 12 tests skipped, 0 scenarios skipped.
```

</details>

With the same config on v1.15.0-pre.0, IPv6 connectivity checks no longer pass (and empirically, no IPv6 connections between pods work).

<details>
<summary>1.15.0 connectivity test report</summary>

```
ℹ️  Cilium version: 1.15.0
🏃 Running tests...
[=] Test [no-policies]
....................
  [-] Scenario [no-policies/pod-to-world]
  [.] Action [no-policies/pod-to-world/http-to-one.one.one.one-0: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> one.one.one.one-http (one.one.one.one:80)]
  [.] Action [no-policies/pod-to-world/https-to-one.one.one.one-0: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> one.one.one.one-https (one.one.one.one:443)]
  [.] Action [no-policies/pod-to-world/https-to-one.one.one.one-index-0: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> one.one.one.one-https-index (one.one.one.one:443)]
  [.] Action [no-policies/pod-to-world/http-to-one.one.one.one-1: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> one.one.one.one-http (one.one.one.one:80)]
  [.] Action [no-policies/pod-to-world/https-to-one.one.one.one-1: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> one.one.one.one-https (one.one.one.one:443)]
  [.] Action [no-policies/pod-to-world/https-to-one.one.one.one-index-1: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> one.one.one.one-https-index (one.one.one.one:443)]
  [-] Scenario [no-policies/pod-to-host]
  [.] Action [no-policies/pod-to-host/ping-ipv4-0: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> 169.229.226.7 (169.229.226.7:0)]
  [.] Action [no-policies/pod-to-host/ping-ipv4-1: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> 169.229.226.9 (169.229.226.9:0)]
  [.] Action [no-policies/pod-to-host/ping-ipv4-2: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> 169.229.226.8 (169.229.226.8:0)]
  [.] Action [no-policies/pod-to-host/ping-ipv4-3: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> 169.229.226.10 (169.229.226.10:0)]
  [.] Action [no-policies/pod-to-host/ping-ipv4-4: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> 169.229.226.7 (169.229.226.7:0)]
  [.] Action [no-policies/pod-to-host/ping-ipv4-5: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> 169.229.226.9 (169.229.226.9:0)]
  [.] Action [no-policies/pod-to-host/ping-ipv4-6: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> 169.229.226.8 (169.229.226.8:0)]
  [.] Action [no-policies/pod-to-host/ping-ipv4-7: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> 169.229.226.10 (169.229.226.10:0)]
  [-] Scenario [no-policies/pod-to-external-workload]
  [-] Scenario [no-policies/pod-to-cidr]
  [.] Action [no-policies/pod-to-cidr/external-1111-0: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> external-1111 (1.1.1.1:443)]
  [.] Action [no-policies/pod-to-cidr/external-1111-1: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> external-1111 (1.1.1.1:443)]
  [.] Action [no-policies/pod-to-cidr/external-1001-0: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> external-1001 (1.0.0.1:443)]
  [.] Action [no-policies/pod-to-cidr/external-1001-1: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> external-1001 (1.0.0.1:443)]
  [-] Scenario [no-policies/pod-to-pod]
  [.] Action [no-policies/pod-to-pod/curl-ipv4-0: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> cilium-test/echo-other-node-5cbfc6f76f-xrvxz (10.244.139.200:8080)]
  [.] Action [no-policies/pod-to-pod/curl-ipv6-0: cilium-test/client-84bfddc76b-zpknp (2607:f140:8801:1::346) -> cilium-test/echo-other-node-5cbfc6f76f-xrvxz (2607:f140:8801:1::222:8080)]
  ❌ command "curl -w %{local_ip}:%{local_port} -> %{remote_ip}:%{remote_port} = %{response_code} --silent --fail --show-error --output /dev/null --connect-timeout 2 --max-time 10 http://[2607:f140:8801:1::222]:8080" failed: command terminated with exit code 28
  ℹ️  curl output:
  
  
  📄 No flows recorded for peer cilium-test/client-84bfddc76b-zpknp during action curl-ipv6-0
  📄 No flows recorded for peer cilium-test/echo-other-node-5cbfc6f76f-xrvxz during action curl-ipv6-0
  [.] Action [no-policies/pod-to-pod/curl-ipv4-1: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> cilium-test/echo-same-node-6fc6fcfb7-9wvtq (10.244.140.252:8080)]
  [.] Action [no-policies/pod-to-pod/curl-ipv6-1: cilium-test/client-84bfddc76b-zpknp (2607:f140:8801:1::346) -> cilium-test/echo-same-node-6fc6fcfb7-9wvtq (2607:f140:8801:1::3d1:8080)]
  ❌ command "curl -w %{local_ip}:%{local_port} -> %{remote_ip}:%{remote_port} = %{response_code} --silent --fail --show-error --output /dev/null --connect-timeout 2 --max-time 10 http://[2607:f140:8801:1::3d1]:8080" failed: command terminated with exit code 7
  ℹ️  curl output:
  
  
  📄 No flows recorded for peer cilium-test/client-84bfddc76b-zpknp during action curl-ipv6-1
  📄 No flows recorded for peer cilium-test/echo-same-node-6fc6fcfb7-9wvtq during action curl-ipv6-1
  [.] Action [no-policies/pod-to-pod/curl-ipv4-2: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> cilium-test/echo-other-node-5cbfc6f76f-xrvxz (10.244.139.200:8080)]
  [.] Action [no-policies/pod-to-pod/curl-ipv6-2: cilium-test/client2-764b565764-js5zx (2607:f140:8801:1::33d) -> cilium-test/echo-other-node-5cbfc6f76f-xrvxz (2607:f140:8801:1::222:8080)]
  ❌ command "curl -w %{local_ip}:%{local_port} -> %{remote_ip}:%{remote_port} = %{response_code} --silent --fail --show-error --output /dev/null --connect-timeout 2 --max-time 10 http://[2607:f140:8801:1::222]:8080" failed: command terminated with exit code 28
  ℹ️  curl output:
  
  
  📄 No flows recorded for peer cilium-test/client2-764b565764-js5zx during action curl-ipv6-2
  📄 No flows recorded for peer cilium-test/echo-other-node-5cbfc6f76f-xrvxz during action curl-ipv6-2
  [.] Action [no-policies/pod-to-pod/curl-ipv4-3: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> cilium-test/echo-same-node-6fc6fcfb7-9wvtq (10.244.140.252:8080)]
  [.] Action [no-policies/pod-to-pod/curl-ipv6-3: cilium-test/client2-764b565764-js5zx (2607:f140:8801:1::33d) -> cilium-test/echo-same-node-6fc6fcfb7-9wvtq (2607:f140:8801:1::3d1:8080)]
  ❌ command "curl -w %{local_ip}:%{local_port} -> %{remote_ip}:%{remote_port} = %{response_code} --silent --fail --show-error --output /dev/null --connect-timeout 2 --max-time 10 http://[2607:f140:8801:1::3d1]:8080" failed: command terminated with exit code 7
  ℹ️  curl output:
  
  
  📄 No flows recorded for peer cilium-test/client2-764b565764-js5zx during action curl-ipv6-3
  📄 No flows recorded for peer cilium-test/echo-same-node-6fc6fcfb7-9wvtq during action curl-ipv6-3
  [-] Scenario [no-policies/client-to-client]
  [.] Action [no-policies/client-to-client/ping-ipv4-0: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> cilium-test/client2-764b565764-js5zx (10.244.140.141:0)]
  [.] Action [no-policies/client-to-client/ping-ipv6-0: cilium-test/client-84bfddc76b-zpknp (2607:f140:8801:1::346) -> cilium-test/client2-764b565764-js5zx (2607:f140:8801:1::33d:0)]
  ❌ command "ping -c 1 -6 -W 2 -w 10 2607:f140:8801:1::33d" failed: command terminated with exit code 1
  ℹ️  ping output:
  
  
  📄 No flows recorded for peer cilium-test/client-84bfddc76b-zpknp during action ping-ipv6-0
  📄 No flows recorded for peer cilium-test/client2-764b565764-js5zx during action ping-ipv6-0
  [.] Action [no-policies/client-to-client/ping-ipv4-1: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> cilium-test/client-84bfddc76b-zpknp (10.244.140.83:0)]
  [.] Action [no-policies/client-to-client/ping-ipv6-1: cilium-test/client2-764b565764-js5zx (2607:f140:8801:1::33d) -> cilium-test/client-84bfddc76b-zpknp (2607:f140:8801:1::346:0)]
  ❌ command "ping -c 1 -6 -W 2 -w 10 2607:f140:8801:1::346" failed: command terminated with exit code 1
  ℹ️  ping output:
  
  
  📄 No flows recorded for peer cilium-test/client2-764b565764-js5zx during action ping-ipv6-1
  📄 No flows recorded for peer cilium-test/client-84bfddc76b-zpknp during action ping-ipv6-1
  [-] Scenario [no-policies/pod-to-service]
  [.] Action [no-policies/pod-to-service/curl-0: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> cilium-test/echo-other-node (echo-other-node:8080)]
  [.] Action [no-policies/pod-to-service/curl-1: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> cilium-test/echo-same-node (echo-same-node:8080)]
  [.] Action [no-policies/pod-to-service/curl-2: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> cilium-test/echo-other-node (echo-other-node:8080)]
  [.] Action [no-policies/pod-to-service/curl-3: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> cilium-test/echo-same-node (echo-same-node:8080)]
  [-] Scenario [no-policies/pod-to-hostport]
  [.] Action [no-policies/pod-to-hostport/curl-0: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> cilium-test/echo-other-node-5cbfc6f76f-xrvxz (169.229.226.7:40000)]
  [.] Action [no-policies/pod-to-hostport/curl-1: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> cilium-test/echo-same-node-6fc6fcfb7-9wvtq (169.229.226.9:40000)]
  [.] Action [no-policies/pod-to-hostport/curl-2: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> cilium-test/echo-other-node-5cbfc6f76f-xrvxz (169.229.226.7:40000)]
  [.] Action [no-policies/pod-to-hostport/curl-3: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> cilium-test/echo-same-node-6fc6fcfb7-9wvtq (169.229.226.9:40000)]
[=] Test [no-policies-extra]
................
[=] Test [allow-all-except-world]
..
  ℹ️  📜 Applying CiliumNetworkPolicy 'allow-all-except-world' to namespace 'cilium-test'..
  [-] Scenario [allow-all-except-world/pod-to-external-workload]
  [-] Scenario [allow-all-except-world/pod-to-pod]
  [.] Action [allow-all-except-world/pod-to-pod/curl-ipv4-0: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> cilium-test/echo-other-node-5cbfc6f76f-xrvxz (10.244.139.200:8080)]
  [.] Action [allow-all-except-world/pod-to-pod/curl-ipv6-0: cilium-test/client-84bfddc76b-zpknp (2607:f140:8801:1::346) -> cilium-test/echo-other-node-5cbfc6f76f-xrvxz (2607:f140:8801:1::222:8080)]
  ❌ command "curl -w %{local_ip}:%{local_port} -> %{remote_ip}:%{remote_port} = %{response_code} --silent --fail --show-error --output /dev/null --connect-timeout 2 --max-time 10 http://[2607:f140:8801:1::222]:8080" failed: command terminated with exit code 28
  ℹ️  curl output:
  
  
  📄 No flows recorded for peer cilium-test/client-84bfddc76b-zpknp during action curl-ipv6-0
  📄 No flows recorded for peer cilium-test/echo-other-node-5cbfc6f76f-xrvxz during action curl-ipv6-0
  [.] Action [allow-all-except-world/pod-to-pod/curl-ipv4-1: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> cilium-test/echo-same-node-6fc6fcfb7-9wvtq (10.244.140.252:8080)]
  [.] Action [allow-all-except-world/pod-to-pod/curl-ipv6-1: cilium-test/client-84bfddc76b-zpknp (2607:f140:8801:1::346) -> cilium-test/echo-same-node-6fc6fcfb7-9wvtq (2607:f140:8801:1::3d1:8080)]
  ❌ command "curl -w %{local_ip}:%{local_port} -> %{remote_ip}:%{remote_port} = %{response_code} --silent --fail --show-error --output /dev/null --connect-timeout 2 --max-time 10 http://[2607:f140:8801:1::3d1]:8080" failed: command terminated with exit code 7
  ℹ️  curl output:
  
  
  📄 No flows recorded for peer cilium-test/client-84bfddc76b-zpknp during action curl-ipv6-1
  📄 No flows recorded for peer cilium-test/echo-same-node-6fc6fcfb7-9wvtq during action curl-ipv6-1
  [.] Action [allow-all-except-world/pod-to-pod/curl-ipv4-2: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> cilium-test/echo-other-node-5cbfc6f76f-xrvxz (10.244.139.200:8080)]
  [.] Action [allow-all-except-world/pod-to-pod/curl-ipv6-2: cilium-test/client2-764b565764-js5zx (2607:f140:8801:1::33d) -> cilium-test/echo-other-node-5cbfc6f76f-xrvxz (2607:f140:8801:1::222:8080)]
  ❌ command "curl -w %{local_ip}:%{local_port} -> %{remote_ip}:%{remote_port} = %{response_code} --silent --fail --show-error --output /dev/null --connect-timeout 2 --max-time 10 http://[2607:f140:8801:1::222]:8080" failed: command terminated with exit code 28
  ℹ️  curl output:
  
  
  📄 No flows recorded for peer cilium-test/client2-764b565764-js5zx during action curl-ipv6-2
  📄 No flows recorded for peer cilium-test/echo-other-node-5cbfc6f76f-xrvxz during action curl-ipv6-2
  [.] Action [allow-all-except-world/pod-to-pod/curl-ipv4-3: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> cilium-test/echo-same-node-6fc6fcfb7-9wvtq (10.244.140.252:8080)]
  [.] Action [allow-all-except-world/pod-to-pod/curl-ipv6-3: cilium-test/client2-764b565764-js5zx (2607:f140:8801:1::33d) -> cilium-test/echo-same-node-6fc6fcfb7-9wvtq (2607:f140:8801:1::3d1:8080)]
  ❌ command "curl -w %{local_ip}:%{local_port} -> %{remote_ip}:%{remote_port} = %{response_code} --silent --fail --show-error --output /dev/null --connect-timeout 2 --max-time 10 http://[2607:f140:8801:1::3d1]:8080" failed: command terminated with exit code 7
  ℹ️  curl output:
  
  
  📄 No flows recorded for peer cilium-test/client2-764b565764-js5zx during action curl-ipv6-3
  📄 No flows recorded for peer cilium-test/echo-same-node-6fc6fcfb7-9wvtq during action curl-ipv6-3
  [-] Scenario [allow-all-except-world/client-to-client]
  [.] Action [allow-all-except-world/client-to-client/ping-ipv4-0: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> cilium-test/client2-764b565764-js5zx (10.244.140.141:0)]
  [.] Action [allow-all-except-world/client-to-client/ping-ipv6-0: cilium-test/client-84bfddc76b-zpknp (2607:f140:8801:1::346) -> cilium-test/client2-764b565764-js5zx (2607:f140:8801:1::33d:0)]
  ❌ command "ping -c 1 -6 -W 2 -w 10 2607:f140:8801:1::33d" failed: command terminated with exit code 1
  ℹ️  ping output:
  
  
  📄 No flows recorded for peer cilium-test/client-84bfddc76b-zpknp during action ping-ipv6-0
  📄 No flows recorded for peer cilium-test/client2-764b565764-js5zx during action ping-ipv6-0
  [.] Action [allow-all-except-world/client-to-client/ping-ipv4-1: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> cilium-test/client-84bfddc76b-zpknp (10.244.140.83:0)]
  [.] Action [allow-all-except-world/client-to-client/ping-ipv6-1: cilium-test/client2-764b565764-js5zx (2607:f140:8801:1::33d) -> cilium-test/client-84bfddc76b-zpknp (2607:f140:8801:1::346:0)]
  ❌ command "ping -c 1 -6 -W 2 -w 10 2607:f140:8801:1::346" failed: command terminated with exit code 1
  ℹ️  ping output:
  
  
  📄 No flows recorded for peer cilium-test/client2-764b565764-js5zx during action ping-ipv6-1
  📄 No flows recorded for peer cilium-test/client-84bfddc76b-zpknp during action ping-ipv6-1
  [-] Scenario [allow-all-except-world/pod-to-service]
  [.] Action [allow-all-except-world/pod-to-service/curl-0: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> cilium-test/echo-other-node (echo-other-node:8080)]
  [.] Action [allow-all-except-world/pod-to-service/curl-1: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> cilium-test/echo-same-node (echo-same-node:8080)]
  [.] Action [allow-all-except-world/pod-to-service/curl-2: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> cilium-test/echo-same-node (echo-same-node:8080)]
  [.] Action [allow-all-except-world/pod-to-service/curl-3: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> cilium-test/echo-other-node (echo-other-node:8080)]
  [-] Scenario [allow-all-except-world/pod-to-host]
  [.] Action [allow-all-except-world/pod-to-host/ping-ipv4-0: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> 169.229.226.7 (169.229.226.7:0)]
  [.] Action [allow-all-except-world/pod-to-host/ping-ipv4-1: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> 169.229.226.9 (169.229.226.9:0)]
  [.] Action [allow-all-except-world/pod-to-host/ping-ipv4-2: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> 169.229.226.8 (169.229.226.8:0)]
  [.] Action [allow-all-except-world/pod-to-host/ping-ipv4-3: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> 169.229.226.10 (169.229.226.10:0)]
  [.] Action [allow-all-except-world/pod-to-host/ping-ipv4-4: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> 169.229.226.10 (169.229.226.10:0)]
  [.] Action [allow-all-except-world/pod-to-host/ping-ipv4-5: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> 169.229.226.7 (169.229.226.7:0)]
  [.] Action [allow-all-except-world/pod-to-host/ping-ipv4-6: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> 169.229.226.9 (169.229.226.9:0)]
  [.] Action [allow-all-except-world/pod-to-host/ping-ipv4-7: cilium-test/client2-764b565764-js5zx (10.244.140.141) -> 169.229.226.8 (169.229.226.8:0)]
  ℹ️  📜 Deleting CiliumNetworkPolicy 'allow-all-except-world' from namespace 'cilium-test'..
[=] Test [client-ingress]
.^CInterrupt received, cancelling tests...
.
  ℹ️  📜 Applying CiliumNetworkPolicy 'client-ingress-from-client2' to namespace 'cilium-test'..
  [-] Scenario [client-ingress/client-to-client]
  [.] Action [client-ingress/client-to-client/ping-ipv4-0: cilium-test/client-84bfddc76b-zpknp (10.244.140.83) -> cilium-test/client2-764b565764-js5zx (10.244.140.141:0)]
  ℹ️  unable to extract exit code from error: context canceled
  [.] Action [client-ingress/client-to-client/ping-ipv6-0: cilium-test/client-84bfddc76b-zpknp (2607:f140:8801:1::346) -> cilium-test/client2-764b565764-js5zx (2607:f140:8801:1::33d:0)]
  🟥 Skipping command execution: context canceled
  ℹ️  📜 Deleting CiliumNetworkPolicy 'client-ingress-from-client2' from namespace 'cilium-test'..
connectivity test failed: context canceled
```

</details>

### Cilium Version

v1.15.0-pre.0

### Kernel Version

Linux adenine 6.3.10 #1-NixOS SMP PREEMPT_DYNAMIC Wed Jun 28 09:14:25 UTC 2023 x86_64 GNU/Linux

### Kubernetes Version

`Server Version: v1.25.3`

### Sysdump

This sysdump is from the broken `v1.15.0-pre.0` install...

https://www.ocf.berkeley.edu/~njha/tmp/cilium-sysdump-20230902-132916.zip

### Relevant log output

```shell
(nothing looks suspicious from the logs i looked at, see sysdump for all logs)
```

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by ti-mo**
Unqualified integers are treated as `int` in C, but the behaviour when doing binary operations on them against unsigned integers is surprising, to say the least.

Consider the following scenario, before this patch:

```
 #define DEFINE_IPV6(NAME, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, a16) \
 DEFINE_U64_I(NAME, 1) = bpf_cpu_to_be64( \
                       (__u64)(a1) << 56 | (__u64)(a2) << 48 | (__u64)(a3) << 40 | \
                       (__u64)(a4) << 32 | (a5) << 24 | (a6) << 16 | (a7) << 8 | (a8)) .. <elip>

 DEFINE_IPV6(ROUTER_IP, 0x0, 0x0, 0x0, 0x0, 0x80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x80, 0x0, 0x0, 0x0);
```

Both a5 and a13 are set to a byte with a most-significant bit of 1. This results in the following output in the .data section during compilation:

```
0000000000000038 <ROUTER_IP_1>:
       7:       ff ff ff ff 80 00 00 00 <unknown>
0000000000000040 <ROUTER_IP_2>:
       8:       ff ff ff ff 80 00 00 00 <unknown>
```

Since the `<< 24` pushed the 1 bit to now be the most-significant bit of an int, equivalent to an int32 on most 64-bit architectures, the subsequent OR operation against a __u64 causes the value to become sign-extended, setting the upper 32 bits of the resulting (unsigned..) integer to 1. Ouch.

This commit treats each member as a u64, and additionally truncates each macro argument to u8 to prevent any of the arguments overlapping as a defensive measure.

Fixes: https://github.com/cilium/cilium/issues/27898

```release-note
Fix wrong host and router IP being used for some IPv6 deployments, which was causing various connectivity problems.
```

### CODE REVIEW

**Review by borkmann — APPROVED** [review]
🚀

File: bpf/lib/static_data.h [hunk 1]
```diff
@@ -28,14 +28,19 @@
 
 #define DEFINE_IPV6(NAME, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10, a11, a12, a13, a14, a15, a16) \
 DEFINE_U64_I(NAME, 1) = bpf_cpu_to_be64( \
-			(__u64)(a1) << 56 | (__u64)(a2) << 48 | (__u64)(a3) << 40 | \
-			(__u64)(a4) << 32 | (a5) << 24 | (a6) << 16 | (a7) << 8 | (a8)); \
+			(__u64)(__u8)(a1) << 56 | (__u64)(__u8)(a2) << 48 | \
+			(__u64)(__u8)(a3) << 40 | (__u64)(__u8)(a4) << 32 | \
+			(__u64)(__u8)(a5) << 24 | (__u64)(__u8)(a6) << 16 | \
+			(__u64)(__u8)(a7) << 8  | (__u64)(__u8)(a8)); \
```
Comment (borkmann) [inline bpf/lib/static_data.h:34]: Is the double cast `(__u64)(__u8)` indeed needed here? Meaning, single one `(__u64)` is not enough?

Comment (ti-mo) [inline bpf/lib/static_data.h:34]: We don't want any values to spill over into the higher bits, so it's okay to be defensive here.

**Comment by borkmann** [comment]
Yes, agree, makes sense! Kudos, amazing catch!

**Pull request merged**